### PR TITLE
fix banner component and support mobile

### PR DESCRIPTION
**Pull request by songgao**
Following the [banner style guide](https://app.zeplin.io/project/56a15d74e1e098dd3116d10b?seid=5c0f21bafd7d8daf22663ec5), and added actions and close button. I'll have a follow-up PR that uses this for Files errors.

image
image

cc @cecileboucheron I might have mis-remembered some of your in-person feedback. Please let me know if something looks off.

### CODE REVIEW

File: shared/common-adapters/text.native.js [hunk 1]
```diff
@@ -130,12 +136,12 @@ function _getStyle(
   type: TextType,
   backgroundMode?: Background = 'Normal',
   lineClampNum?: ?number,
-  clickable?: ?boolean
+  clickable?: ?boolean,
+  forceUnderline: boolean
 ) {
-  if (backgroundMode === 'Normal') return null
```
Comment (songgao) [inline shared/common-adapters/text.native.js]: @chrisnojima Is there a reason why we disabled underline on mobile? I removed this so the underline below would work.

Comment (songgao) [reply shared/common-adapters/text.native.js]: Related: I think with this change, on mobile we'll start underlining links, while previously we didn't automatically underline links on mobile.

Comment (songgao) [reply shared/common-adapters/text.native.js]: I ended up making it not automatically underline links; just when `underline={true}` is provided.

File: shared/common-adapters/banner.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  actions?: Array<{
+    title: string,
+    onClick: () => void,
+  }>,
+  color: Color,
+  onClose?: () => void,
+  text: string,
+}
+
+const getContainerStyle = memoize((props: Props) =>
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: we shouldn't memoize common adapters like this. you'll have no control over your cache getting thrashed

Comment (chrisnojima) [reply shared/common-adapters/banner.js]: if you want to do this (not that i think this is expensive so we should actually not do this at all) but if you did just make a map of the color to the value and be done

Comment (chrisnojima) [reply shared/common-adapters/banner.js]: and in general this is not a good idea to pass props to memoize, esp if you just pulling out one value like color

File: shared/common-adapters/banner.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  Styles.collapseStyles([styles.container, colorToBackgroundColorStyles[props.color]])
+)
+const getTextStyle = memoize((props: Props) => colorToTextColorStyles[props.color])
+
+const Banner = (props: Props) => (
+  <Box2 direction="horizontal" fullWidth={true} style={getContainerStyle(props)}>
+    <Box2 key="textBox" direction="horizontal" style={styles.textContainer} centerChildren={true}>
+      {props.text.split(' ').map((word, index) => (
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: why do this

File: shared/common-adapters/banner.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        <Text key={`word-${index}`} type="BodySmallSemibold" style={getTextStyle(props)}>
+          {word}&nbsp;
+        </Text>
+      ))}
+      {!!props.actions &&
+        props.actions.map(({title, onClick}, index) => (
+          <React.Fragment key={String(index)}>
+            <Text
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: just use a nested text and get rid of the fragment

File: shared/common-adapters/banner.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  },
+  iconContainer: {
+    alignSelf: 'flex-start',
+    height: 40,
+  },
+  textContainer: Styles.platformStyles({
+    common: {
+      flex: 1,
+      flexWrap: 'wrap',
+      paddingBottom: 8,
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: globalMargins

File: shared/common-adapters/banner.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      paddingTop: 8,
+    },
+    isElectron: {
+      paddingLeft: 64,
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: globalMargins

File: shared/common-adapters/banner.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      paddingRight: 64,
+    },
+    isMobile: {
+      paddingLeft: 24,
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: globalMargins

File: shared/common-adapters/banner.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+          onClick={props.onClose}
+        />
+      </Box2>
+    )}
+  </Box2>
+)
+
+const styles = Styles.styleSheetCreate({
+  container: {
+    minHeight: 40,
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: globalMargins

File: shared/common-adapters/banner.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  },
+  icon: {
+    padding: Styles.globalMargins.tiny,
+  },
+  iconContainer: {
+    alignSelf: 'flex-start',
+    height: 40,
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: globalMargins

File: shared/common-adapters/banner.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            <Text key="space" type="BodySmallSemibold">
+              &nbsp;
+            </Text>
+          </React.Fragment>
+        ))}
+    </Box2>
+    {!!props.onClose && (
+      <Box2 key="iconBox" direction="vertical" style={styles.iconContainer} centerChildren={true}>
+        <Icon
+          fontSize={Styles.isMobile ? undefined : 12}
```
Comment (chrisnojima) [inline shared/common-adapters/banner.js]: globalMargins

**Comment by cecileboucheron** [comment]
@songgao Links should be underlined on mobile.

File: shared/common-adapters/banner.js (hunk 4 above)
Comment (chrisnojima) [reply shared/common-adapters/banner.js]: you can nest text. but only text. so no non-text inside of text

Comment (chrisnojima) [reply shared/common-adapters/banner.js]: https://facebook.github.io/react-native/docs/text#nested-text

**Comment by songgao** [comment]
@chrisnojima Thanks fro the review; ready for another look!

**Comment by songgao** [comment]
bump!

**Comment by songgao** [comment]
bump!

File: shared/common-adapters/banner.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+const colorToBackgroundColorStyles = {
+  blue: {backgroundColor: Styles.globalColors.blue},
+  green: {backgroundColor: Styles.globalColors.green},
+  grey: {backgroundColor: Styles.globalColors.lightGrey2},
+  red: {backgroundColor: Styles.globalColors.red},
+  yellow: {backgroundColor: Styles.globalColors.yellow},
+}
+
+const colorToTextColorStyles = {
```
Comment (buoyad) [inline shared/common-adapters/banner.js]: You could use `styleSheetCreate` for this and `colorToBackground` to reap the native rewards

**Pull request merged**